src/transport.rs

```diff
@@ -0,0 +1,251 @@
+}
+
+impl EccTransport {
+    pub fn from_path(path: &str, address: u16) -> Result<Self> {
+        
+        if path.starts_with("/dev/tty"){
+            let swi_port = serialport::new(path, SWI_DEFAULT_BAUDRATE)
+            .data_bits(serialport::DataBits::Seven)
+            .parity(serialport::Parity::None)
+            .stop_bits(serialport::StopBits::One)
+            .timeout(Duration::from_millis(50))
+            .open().unwrap_or_else(|e| {
+                eprintln!("Failed to open serial port. Error: {}", e);
+                ::std::process::exit(1);
+            });
+            
+            Ok(Self {
+                i2c_port: None,
+                i2c_address: address,
+                swi_port: Some(swi_port),
+                protocol: TransportProtocol::Swi
+            })    
+        }
+        else if path.starts_with("/dev/i2c") {
+            let mut i2c = I2c::from_path(path)?;
+            i2c.smbus_set_slave_address(address, false)?;
+            
+            Ok(Self { 
+                swi_port: None,
+                i2c_port: Some(i2c),
+                i2c_address: address,
+                protocol: TransportProtocol::I2c
+            })
+        }
+        else {
+            eprintln!("Failed to open selected port");
+            ::std::process::exit(1);
+        }
+    }
+
+    pub fn send_wake(&mut self) -> Result {
+        match self.protocol {
+            TransportProtocol::I2c =>{
+                let _ = self.send_i2c_buf(0, &[0]);
+                thread::sleep(WAKE_DELAY);
+                Ok(())
+            }
+            TransportProtocol::Swi => {
+                if let Err(_err) = self.swi_port.as_mut().unwrap().set_baud_rate(SWI_WAKE_BAUDRATE)
+                {
+                    return Err(Error::timeout());
+                }
+                
+                let _ = self.swi_port.as_mut().unwrap().write(&[0]);
+                
+                thread::sleep(WAKE_DELAY);
+                let _ = self.swi_port.as_mut().unwrap().set_baud_rate(SWI_DEFAULT_BAUDRATE);
+                let _ = self.swi_port.as_mut().unwrap().clear(ClearBuffer::All);
+                Ok(()) 
+            },
+        }
+    }
+
+    pub fn send_sleep(&mut self) {
+        match self.protocol {
+            TransportProtocol::I2c => {
+                let _ = self.send_i2c_buf(self.i2c_address, &[1]);
+            } 
+            TransportProtocol::Swi => {
+                let mut sleep_msg = BytesMut::new();
+                sleep_msg.put_u8(ATCA_SWI_SLEEP_FLAG);
+                let sleep_encoded = self.encode_uart_to_swi(&sleep_msg);
+        
+                let _ = self.swi_port.as_mut().unwrap().write(&sleep_encoded);
+                thread::sleep( SWI_BIT_SEND_DELAY * 8);
+            },
+        }
+    }
+
+    pub fn send_recv_buf(&mut self, delay: Duration, buf: &mut BytesMut) -> Result {
+        match self.protocol {
+            TransportProtocol::I2c => {
```

Comment:
Implement these on a `TransportProtocol` instead of matching on protocol all over the place?

Reply:
Could you elaborate on how this would be done? I'm unsure of how implementing on the enum would get away from matching? Wouldn't you still have to match within the `TransportProtocol` implementation?